### PR TITLE
Attr values can be any kind of enum

### LINKED ISSUE #2256: Support range values on integer attrs
I would like to declare a component attr as such:

```
  attr :value, :integer,
    doc: "A progress value, between 0 and 100.",
    values: 0..100
```

In the storybook library, we're currently handling range values with a dedicated number input

### PR DESCRIPTION
**Pull request by cblavier**
Resolves #2256

### CODE REVIEW

File: lib/phoenix_component/declarative.ex [hunk 1]
```diff
@@ -436,7 +436,7 @@ defmodule Phoenix.Component.Declarative do
   end
 
   defp validate_attr_values!(slot, name, type, values, line, file) do
-    unless is_list(values) and not Enum.empty?(values) do
+    unless is_enumerable(values) and not Enum.empty?(values) do
       compile_error!(line, file, """
       :values must be a non-empty list, got: #{inspect(values)}
```
Comment (josevalim) [inline lib/phoenix_component/declarative.ex:441]:
```suggestion
      :values must be a non-empty enumerable, got: #{inspect(values)}
```

File: lib/phoenix_component/declarative.ex [hunk 2]
```diff
@@ -447,6 +447,10 @@ defmodule Phoenix.Component.Declarative do
         do: bad_value!(slot, name, type, value, line, file)
   end
 
+  defp is_enumerable(values) do
+    !is_nil(Enumerable.impl_for(values))
```
Comment (josevalim) [inline lib/phoenix_component/declarative.ex:451]:
```suggestion
    Enumerable.impl_for(values) != nil
```

File: test/phoenix_component/declarative_assigns_test.exs [hunk 3]
```diff
@@ -1310,6 +1310,28 @@ defmodule Phoenix.ComponentDeclarativeAssignsTest do
...
+
+    assert_raise CompileError, msg, fn ->
+      defmodule Phoenix.ComponentTest.AttrDefaultValuesMismatch do
+        use Elixir.Phoenix.Component
+
+        attr :foo, :integer, default: 11, values: 1..10
+        def func(assigns), do: ~H[]
+      end
+    end
+  end
```
Comment (josevalim) [inline test/phoenix_component/declarative_assigns_test.exs:1333]: I think we should also add tests that the validation happens as desired?

Comment (cblavier) [reply test/phoenix_component/declarative_assigns_test.exs:1333]: you mean a test with a default value within the values range?

Comment (josevalim) [reply test/phoenix_component/declarative_assigns_test.exs:1333]: That a value within the range does not warn and a value outside the range warns. :)

Comment (cblavier) [reply test/phoenix_component/declarative_assigns_test.exs:1333]: got it, it's `verify_test.exs` I should amend

Comment (cblavier) [reply test/phoenix_component/declarative_assigns_test.exs:1333]: Done. ✅ 
Also improved tests to count how many warnings were emitted and ensure that good values did not raise

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**